subprojects/file-watching/src/main/java/org/gradle/internal/watch/registry/FileWatcherUpdater.java

```diff
@@ -24,6 +25,51 @@
 import java.io.File;
 import java.util.Collection;
 
+/**
+ * <p>
+ * Responsible for updating the file watchers based on changes to the contents of the virtual file system,
+ * and changes to the file system hierarchies we are allowed to watch.</p>
+ *
+ * <p>
+ * The following terms are worth distinguishing between:
+ * </p>
+ *
+ * <dl>
+ *     <dt>file system hierarchy</dt>
+ *     <dd>A directory on the file system with all its descendants.</dd>
+ *
+ *     <dt>watchable hierarchies</dt>
+ *     <dd>The list of file system hierarchies Gradle might want to watch if interesting content appears in the VFS for them.
+ *     These are currently roots of the builds seen by the daemon during the current or previous invocations.
+ *     Interesting content means that we keep track of something that actually exists under the hierarchy (so not exclusively missing files).
+ *     Note that hierarchies can be nested in each other.</dd>
+ *
+ *     <dt>recently used watchable hierarchies</dt>
+ *     <dd>The list of file system hierarchies Gradle has accessed recently. These include build roots from recent builds
+ *     that we actually accessed.</dd>
+ *
+ *     <dt>watched files</dt>
+ *     <dd>A {@link org.gradle.internal.file.FileHierarchySet} of the files that we are currently watching.
+ *     This helps decide whether or not something is being watched in a quick way.</dd>
+ *
+ *     <dt>watched hierarchies</dt>
```

Comment:
We only need this for the hierarchical watchers. Then the doc here needs to change a little bit as well.

We said we'd move this to hierarchical watcher updater in a separate PR.